### ISSUE TITLE
Feat: Support child spaces in WebDAV view

### ISSUE THREAD

**Issue by grignards** [tracim/tracim#3549]
## Feature description and goals

The child spaces (see #3546) introduce new levels of "folders" in WebDAV, we have to take care those work properly:

- [x] the sub-spaces shall be displayed as folders (check if it is possible to distinguish them from folders which are contents)
- [x] check that sub-space move is properly disabled
- [x] sub-spaces shall have a `.space` extension to avoid name collision with folders

## Prerequisites

- #3546
- #3625

### How do we handle the possibility of having a folder+sub-space with the same name ?

Seen 4 possibilities:

- forbid having same name between folder and space.
- ~~add a suffix in case of conflict (for example `~s<workspace_id>` and `~f<content_id>`): a bit like space naming conflict (see #2477)~~ deprecated as we do decide to remove the support for same id space, see #3625 
- add an extension to spaces (`.space`): easy but could disturb users
- re-organize spaces in webdav in order to separate sub-spaces and folders. This would require a reserved space name (eg. `All Contents`).

**Decision**: we'll go with the `.space` extension.

## Implemented solution

https://github.com/tracim/tracim/pull/3687

Features:
- add filemanager_filename property for workspace to access to webdav showed version of workspace and use it in order that all workspace end with this extension in webdav view
- adapt workspace to be able to show both subworkspace and content, workspace have priority so if both workspace and folder exist with same webdav name (if you create dir with .space extension) it means workspace is prioritary and you will not be able to see the folder. It' s a similar issue as same workspace name for one user.
- adapt workspace to show orphan workspace (workspace with parent not accessible) at root.

Refactoring/Tests:
- remove some unused code like get_folder_with_workspace_path_labels or filter_query_for_content_label_as_path methods
- adapt test with the new space extension.
- add WebdavPath object helping process path hierarchy for webdav instead of old static method.
- add few tests to check if children and orphans workspaces are accessible as expected.

- Refactor Webdav ressource in order to avoid confusing inheritance by using composition with WebdavContainer class.

**Comment by inkhey** [comment]
just found 2 things:
- creating a folder in space will create a folder.
- we can have issue with same name for folder and subspace. in fact we do allow same name space as far i know. This can cause issue in some file manager

**Comment by inkhey** [comment]
> just found 2 things:
> 
>     * creating a folder in space will create a folder.
> 
>     * we can have issue with same name for folder and subspace. in fact we do allow same name space as far i know. This can cause issue in some file manager

In fact, this will cause issue, that mean some dir/workspace will not be accessible through webdav as the directory name is in the path.

**Comment by grignards** [comment]
> In fact, this will cause issue, that mean some dir/workspace will not be accessible through webdav as the directory name is in the path.

Yes, I guess that forbidding to give the same name to a sub-space and a folder is logical and not too restrictive for users.

**Comment by inkhey** [comment]
> > In fact, this will cause issue, that mean some dir/workspace will not be accessible through webdav as the directory name is in the path.
> 
> Yes, I guess that forbidding to give the same name to a sub-space and a folder is logical and not too restrictive for users.

i just retested for spaces, as multiple same name space are allowed (because as space are not visible by anyone, it's will be very annoying to force an unique name here).
There does exist a workaround to make the mechanism mostly work on webdav (doesn't allow renaming all workspace):
https://github.com/tracim/tracim/issues/2477

![Capture du 2020-09-14 14-53-30](https://user-images.githubusercontent.com/7851652/93088357-1838e600-f69a-11ea-9964-0706f8bfdc2b.png)

**Comment by inkhey** [comment]
2 idea about workaround to avoid putting new rules:
- distinctive name for space: all space follow a similar pattern, naive idea is to use emoji (not sure it's a good idea but this is visually less awful if Unicode is correctly handled: `🗄️mon espace` instead of `mon espace`. This way it should be possible to keep the rename mechanism working for first visible same name workspace.
- Reorganize the webdav view in order to not show folder and workspace at the same level , this always mean potential collision but we can decide a reserved name for this, this create also an issue about internationalization:

suggestion (reserved name is `🗄️ my data`)
```
my space
  \-- my space
     \-- 🗄️ my data
     \- my subspace
       \-- 🗄️ my data
  \-- my second subspace
     \-- 🗄️ my data
  \-- 🗄️ my data
```

**Comment by grignards** [comment]
In order to minimize the modifications and impacts I would go for an extension of the current behavior when two spaces have the same name. If a space has the same name as a folder, add `~~<workspace_id>` to its name.

**Comment by raphj** [comment]
Does this mean that the path of a space can change depending on whether a new folder or file has the same name, or stops having the same name?

If we are okay with this drawback, :rocket:

**Comment by raphj** [comment]
Should we unconditionally allow accessing a space with the `~~<workspace_id>` suffix, even without collision, even if we don't list it in the parent to mitigate this drawback?

**Comment by grignards** [comment]
> Should we unconditionally allow accessing a space with the `~~<workspace_id>` suffix, even without collision, even if we don't list it in the parent to mitigate this drawback?

This is possibly already the case, @inkhey?

**Comment by inkhey** [comment]
> In order to minimize the modifications and impacts I would go for an extension of the current behavior when two spaces have the same name. If a space has the same name as a folder, add `~~<workspace_id>` to its name.

The problem with using this kind of naming is that if we use it the way we currently do, it disable two feature. I'm ok to remove them, but this need to be clarified:
- create workspace
- renaming workspace

In current implementation we can only rename the workspace without "~" (first one with same name) and we are able to create workspace. I don't see any way to activate workspace creation if we do use ~~ naming as many client webdav will bug due to "i created a dir but i don't see it" bugs.

I'm also not a huge fan of the "~~" syntax, so i'm ok with a better syntax. The problem i do see with "~~" here is that's it's to me not really clear from user point of view, that's it's about space/workspace. Note that's If we do disallow workspace creation/renaming, we can decide that naming is something like "blabla.space.12" and "blabla.espace.12" with different naming for different language (according to user lang) without any trouble.

> > Should we unconditionally allow accessing a space with the `~~<workspace_id>` suffix, even without collision, even if we don't list it in the parent to mitigate this drawback?
> 
> This is possibly already the case, @inkhey?

In current implementation, the idea is that if we have a name with "~~" at the end, it's will get id from here instead of relying on name. So, the path to the first workspace already works with "~~" and it's not a big deal to make it default behavior. As i said, the problem is more about support for renaming/creating space.

**Comment by lebouquetin** [comment]
I think we need a simple mechanism. So I suggest:

- access to spaces is read-only through webdav
- spaces are named `<label>.space`
- it is acceptable to forbid folders with a name like `something.space`

Adding tilds and numeric ids is counter intuitive for common users

**Comment by lebouquetin** [comment]
In a synchronisation clients, spaces wont be creatable through the client: we'll check a list of spaces to synchronize and we'll synchronize them in choosen folders, like it's done with git for example.

**Comment by inkhey** [comment]
> I think we need a simple mechanism. So I suggest:
> 
>     * access to spaces is read-only through webdav
> 
>     * spaces are named `<label>.space`
> 
>     * it is acceptable to forbid folders with a name like `something.space`
> 
> 
> Adding tilds and numeric ids is counter intuitive for common users

This does not answers the same naming space case.

**Comment by inkhey** [comment]
> > I think we need a simple mechanism. So I suggest:
> > ```
> > * access to spaces is read-only through webdav
> > 
> > * spaces are named `<label>.space`
> > 
> > * it is acceptable to forbid folders with a name like `something.space`
> > ```
> > 
> > 
> > Adding tilds and numeric ids is counter intuitive for common users
> 
> This does not answers the same naming space case.

I suggest either:

- <label>.<space_id>.space
- <label>.space.<space_id>

and either unconditionally show workspace_id or show it only when needed (as current behavior with `~~`) .

**Comment by inkhey** [comment]
as explian here https://github.com/tracim/tracim/issues/3602#issuecomment-696766781, disabling workspace creating from webdav may solve another problems too.

**Comment by lebouquetin** [comment]
> I suggest either <label>.<space_id>.space or <label>.space.<space_id> and

we do not need to put <space_id> because the user will never face this case in webdav only. In the user interface the user will be in the impossibility to choose the right space. So this  is the same for webdav. So its acceptable not to help her/him

**Comment by grignards** [comment]
> as explained here [#3602 (comment)](https://github.com/tracim/tracim/issues/3602#issuecomment-696766781), disabling workspace creating from webdav may solve another problems too.

I agree that disabling workspace creation through webdav would simplify things and as @lebouquetin also agrees I think we can do this. @inkhey Could you write an issue for this? I do not know if creating spaces with webdav is documented/advertised to our users, but we should add a task/comment in the "prepare 3.2.0 release" issue to complete the changelog with this change.

**Comment by inkhey** [comment]
@grignards are you ok, to go for the `.space` naming principle ?
- forbid having same name between folder and space. -> complex and may cause some trouble.
- re-organize spaces in webdav in order to separate sub-spaces and folders. −> possible but may create lot of directory.
- `.space` -> not so pretty.

If yes, the only remaining points i see here is : 
- is `.space` ok ?
- should we internationalize `.space` ?

The internationalization point can be solve later and renaming .space should also been not a big deal so we can create another ticket if we are not sure about these point, and change the naming later with a small patch.

**Comment by grignards** [comment]
> @grignards are you ok, to go for the `.space` naming principle ?
> 
>     * forbid having same name between folder and space. -> complex and may cause some trouble.
> 
>     * re-organize spaces in webdav in order to separate sub-spaces and folders. −> possible but may create lot of directory.
> 
>     * `.space` -> not so pretty.
> 
> 
> If yes, the only remaining points i see here is :
> 
>     * is `.space` ok ?

For me yes.

>     * should we internationalize `.space` ?

Not right now, I'm not even sure anybody will ever have one folder and one sub-space with the same name…

> 
> The internationalization point can be solve later and renaming .space should also been not a big deal so we can create another ticket if we are not sure about these point, and change the naming later with a small patch.

**Comment by inkhey** [comment]
This need https://github.com/tracim/tracim/issues/3546 to be fixed before starting this one.
One point unclear with this one is about the rule to avoid collision: 
- We should unallow both file and directory ending with` .space`, it is acceptable ?
- How can we handle in this case the potential previous file ending with .space ?

As far i see, from what i see in code:
- i do suggest to have specific error code for validating filename each time this is needed.
- this code should be runned at the same time we check for filename already used "case".
- If we do this way a content ending with ".space" will be "broken" until we do rename it.
- maybe tel frontend about this rule, this way we avoid troubles
- if we need another time to be more strict, we do need to complete a bit the backend (and potentially frontend code) code about allowed content filename.

**Comment by lebouquetin** [comment]
@inkhey at the moment, conflicts between files/foldes ending with .space seems to be an acceptable known issue. What is the behavior in case this type of naming is used?

**Comment by inkhey** [comment]
> @inkhey at the moment, conflicts between files/foldes ending with .space seems to be an acceptable known issue. What is the behavior of tracim in case we use this type of naming?

There is actually no rule for this, as the feature is not yet implemented.

**Comment by inkhey** [comment]
> > @inkhey at the moment, conflicts between files/foldes ending with .space seems to be an acceptable known issue. What is the behavior of tracim in case we use this type of naming?
> 
> There is actually no rule for this, as the feature is not yet implemented.

Note: the conflict is between space and all file type including folder. i suggest space are showed and same file/folder with ".space" extension are not visible in this particular case.

**Comment by inkhey** [comment]
> > > @inkhey at the moment, conflicts between files/foldes ending with .space seems to be an acceptable known issue. What is the behavior of tracim in case we use this type of naming?
> > 
> > 
> > There is actually no rule for this, as the feature is not yet implemented.
> 
> Note: the conflict is between space and all file type including folder. i suggest space are showed and same file/folder with ".space" extension are not visible in this particular case.

This is the rule currently implemented is #3676

**Comment by Skylsmoi** [comment]
I suggest waiting for #3579 to test this fully, as we can:

    check consistency of workspace view between webdav an frontend.
    easily create hierarchy of workspace without setting value in database to test

All workspaces features should be tested as this PR does big change in the webdav process.
The new things to test:

    test sub-workspace access
    tests move/copy content between hierarchical workspace.
    test orphans workspaces access (workspace with parent where user is not members)
    test that all workspace end with .space
    test workspace over content priority if same name.

**Comment by raphj** [comment]
Tested on banch, all the cases given in the previous comment.

**Comment by PhilippeAccorsi** [comment]
Done
